syntax = "proto3";
package chat;

message ClientMessage 
{
    oneof payload 
    {
        LoginRequest login = 1;
        ChatRequest global_chat = 2;
        PrivateChatRequest private_chat = 3;
        JoinRoomRequest join_room = 4;
        RoomChatRequest room_chat = 5;
    }
}

message ServerMessage
{
    oneof payload 
    {
        LoginResponse login_response = 1;
        ChatBroadcast chat_broadcast = 2;
        PrivateChatIncoming private_chat = 3;
        ErrorResponse error = 4;
    }
}

message LoginRequest 
{
    string username = 1;
}

message ChatRequest
{
    string text = 1;
}

message PrivateChatRequest
{
    string target_username = 1;
    string text = 2;
}

message JoinRoomRequest
{
    string room_name = 1;
}

message RoomChatRequest
{
    string room_name = 1;
    string text = 2;
}

message LoginResponse
{
    bool success = 1;
    string message = 2;
}

message ChatBroadcast
{
    string sender = 1;
    string text = 2;
    string origin_room = 3;
    uint64 timestamp = 4;
}

message PrivateChatIncoming
{
    string sender = 1;
    string text = 2;
    uint64 timestamp = 3;
}

message ErrorResponse
{
    string reason = 1;
}